src/coreclr/vm/jithelpers.cpp

```diff
@@ -600,16 +607,30 @@ HCIMPL1_V(UINT64, JIT_Dbl2ULng, double val)
 {
     FCALL_CONTRACT;
 
-    const double two63  = 2147483648.0 * 4294967296.0;
-    UINT64 ret;
-    if (val < two63) {
-        ret = FastDbl2Lng(val);
-    }
-    else {
-        // subtract 0x8000000000000000, do the convert then add it back again
-        ret = FastDbl2Lng(val - two63) + I64(0x8000000000000000);
+    // ** NOTE **
+    // This should be kept in sync with FloatingPointUtils::convertDoubleToUInt64
+    // ** NOTE **
+
+    // Previous versions of compilers have had incorrect implementations here, however
+    // all currently supported compiler implementations are believed to be correct.
+
+    if (val <= -1)
+    {
+        // While the Ecma spec does not specifically call this out,
+        // the case of conversion from negative double to unsigned integer is
+        // effectively an overflow and therefore the result is unspecified.
+        // With MSVC for x86/x64, such a conversion results in the bit-equivalent
+        // unsigned value of the conversion to integer. Other compilers convert
+        // negative doubles to zero when the target is unsigned.
+        // To make the behavior consistent across OS's on TARGET_XARCH,
+        // this double cast is needed to conform MSVC behavior.
```

Comment:
Is conforming to the MSVC behavior the right meaningful choice here?

Reply:
There's no correct answer here and I don't have a strong preference on what we do, as long as its consistent.

Notably people are more likely to be dependent on the MSVC behavior than not given its prevalence in .NET so far; although that by itself isn't probably worth being the only deciding factor.

-------------------

The things we should likely consider are:
* For integers, `(ulong)(-1)` is bitwise and so it returns `18446744073709551615`; the MSVC behavior effectively "matches" this
* C# decided to constant fold out of range casts to `0`. Clang/GCC, according to the comment, do the same (I've not validated that yet)
* The underlying x86/x64 instructions return `80000000` and `80000000_00000000` for out of range casts to an integer type
* The underlying ARM64 instructions look to use saturation/clamping (`x <= -1` goes to `0`, `x >= 18446744073709551616.0` goes to `18446744073709551615`)

We won't be able to get perfect consistency here in all cases (at least not without the risk of hurting perf in ML or other "important" scenarios).
Likewise, this gets slightly interesting when comparing scalar vs vector operations.

The simplest and "most consistent" thing would probably be to normalize constant folding for R2R/AOT scenarios and to match what Clang, GCC, and Roslyn do (normalize to zero).
We could then, for JIT time constant folding and JIT time computation, match what the underlying platform does.

If we do both, we then at least get Roslyn/AOT consistency and ensure that users can rely on scalar vs vectorized code being consistent as well.

Reply:
> The simplest and "most consistent" thing would probably be to normalize constant folding for R2R/AOT scenarios and to match what Clang, GCC, and Roslyn do (normalize to zero).

Yes, that sounds like the most sensible thing to do.

> We could then, for JIT time constant folding and JIT time computation, match what the underlying platform does.

Do you mean that we would do something else if the code is JITed? It does not sound right to me.

Reply:
> Do you mean that we would do something else if the code is JITed? It does not sound right to me.

Possibly.

I do not believe we can or should try to normalize floating-point casts in all cases. Doing so would force every `floating-point -> integer` conversion, which for all of ARM64 and most of x64 is currently a single inline instruction, to become a call instead with at least 1-2 branches + the instruction.
I expect this would incur measurable and likely detrimental overhead to ML, graphics, and other floating-point heavy code (including scenarios like WPF).
Likewise, this would cause issues between `Vector<T>`, `Vector64<T>`, `Vector128<T>`, and `Vector256<T>` compared to just the underlying `T` based on whether acceleration is enabled or disabled.

Because of that, I believe the correct thing to do here is for the JIT time helper functions to match the platform behavior where possible for non-constants. This ensures we can keep `floating-point <-> integer` conversions fast and generally consistent.

That then leaves an open area around constant folding. Roslyn does not actually target a specific platform and so they had to pick their own behavior for determinism purposes (which happened to be "fold to zero").

R2R/AOT/etc do target a specific platform, as does constant folding. So we could decide to match Roslyn and return `0` or match the rest of the JIT and the underlying platform. The former would introduce issues where optimizations being enabled vs disabled would change the output of the program. The latter would result in inconsistencies compared to what Roslyn does.

Reply:
I think the right behavior is to be consistent between platforms and compilation modes by default as much as possible for trivial floating point math like casts.

Is it correct that clang normalizes by default for C/C++ everywhere? If clang can afford to pay for this consistency, I think we should pay for it too.

Looking at the arm64 disassembly, the performance seems to be a problem on x64 only. arm64 can do it in a single instruction. Do you agree?

I would be ok with having some inconsistencies for vector math if it is warranted for performance.

Reply:
> I think the right behavior is to be consistent between platforms and compilation modes by default as much as possible for trivial floating point math like casts.

I generally agree. The issue is when that behavior introduces serious performance regressions, as it would in this case.

> Is it correct that clang normalizes by default for C/C++ everywhere?

Clang only normalizes for constant folding. For unknown inputs, there is no normalization and you get the behavior of the underlying platform.

You can see this here [full link to godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGISdKuADJ4DJgAcj4ARpjEEhqkAA6oCoRODB7evv7SyamOAiFhkSwxcVwJdpgO6UIETMQEmT5%2BAbaY9gUMdQ0ERRHRsfG29Y3N2W0Ko32hA6VDFQCUtqhexMjsHOYAzKHI3lgA1CbbblP4ggB0CCfYJhoAgvcPXqEEAGySAPoEh2HATEcADdMF90KsovQfqgvq9BJ8IItniYAOxWR6HTGHYiYAhrBiHCBwj7fAiLAC0XEuGhO6KeKIAIsjHsTPj9DgZiMBQeCvJDQQJQURYW8EUjHqi6Vjsbj8YTWaTFlwABwBd4ogIaFHbFEaACcAFYDVx3ibqbTkYzmQ8qLRUIDDnhtmZoV8qM6IG9nezVOL6VKsTi8cQCaoLRKrRLHrz%2BY7vcKqAivS7fr7LQHMUG5WHtnTUUyoza7Q68GyEx7RaTDmmIxmZcHQ%2BH6QWntGIfRHWWYYnJJ74VWa/7ntKsyHq0389bk%2Bz3SmYU6zNRi6m/ZLh4HZWOc3nI62HpWZ/H5wjbfaV%2Bn15nN43c5aW89ifHDrPXV4PaeHYO1xiNw3x7eI3vFkD1%2BF9hS8E9l2rVc0UvetswnXdnmnUCuy%2BBcIBjDsv1gn8rz/bc7ynfsZzQ0teywzBoIvPD4K3RCgJeFDnzQt9F0o6jazg0cbx3RiFVI0kYQgij2yonC6x4/8%2BOtD9fjY11ZyJZiJO469pKIwsOIUhMEUfFNOKHWipMIwDZKgkTFI9ATzy44z1NM5trW0tCe2UkjbKMh4RwchiOGWWhOANXg/A4LRSFQTg3GsaxDgUVZ1ionYeFIAhNH85YAGsQGdal3jMd53m2FUDR1ZUDT1fROEkEL0oizheAUEAEjSsL/NIOBYBgRAUFQFhEjoWJyEoNB%2BsGuJ9kMYAfmILwGEyvg6AIWImogKI6qiUIGgAT04FLRrYQQAHkGFoXa2tILAWCm8QLvwHEahBJqLswVRqi8Za9t4N4Ojq2g8CiYgdo8LAvtS4g8BYL7lltJhgAUAA1PBMAAdyOxJGDB/hBBEMR2CkGRBEUFR1Au3QzH0KaUBiyx9ABprIGWVBEi6Z7ySO7ZGo6aouhcBh3E8Fo9GCWYSjKPQ8jSARxj8LgkhSKWGH6MWFnaTpammGW9CqGoBB6RplcGcoRl6LW5amXpDfmcplnitYNj0AgcU2HgAqC2qLsijhVGVd5yU%2BQ5JqMQknbmzLFkJaLLGsUhDlwQgSGOMxitjjwxvoYgk%2BdRZeFarRFmWBBMCYLA4kRUhsrMZVLj1MwuD1SQNUkPUUXeZUuANZUqo4GrSChyRq64Kvk4NVuCoCbZCtIULwq9xrmtS9Llk6nrRoGjPhogNfxpABoWCBZVySD4AuAqOWaFoZbiFW9aLs25hiHO/a%2BsOggTrOuqrpuzZwvunm8CenVV671PrcG%2BoIX6F1/qA2BhgF2ucIZQzATDAw8Mkao3RpjMBhNhCiHEATbG8glBqDqroOWBgjDU2jrTaBDNy7M1ZpwdmnM1b/2cBAVwZtSAi2KEbCWCsuhcMll0K24tzbc11t0TWgtsjiPVnraYojVYWzGDI2WJsDaiz4VwW2CUHZyydpgeBbsODBWnnVL2e8D7kiYIHChwBDin2pJcLgkcaY2DjvgIgmdkqpz6uvWIWcdG5yXoXYupdKBZX8O8akKJlQaANGYI0Ghk4qmkIFHuHtZ4NVsAvPO7UV4QCQNvDeFAt7%2BJ3kCZAiREhfCBPXL4BhlpTC%2BKoT4i1L4rUoLfcK98dpgwOowN%2Bp1zq/0wNdIwt1f54AesCTAz1wrAOQB9eB5AIEZPCtAoGj8QarKdpDaGfBUGI2RmjDGoUUqENxvg6QhDiYkLJjlSmlDTDUJsLQ%2BATMWbpDZkdMwXN5F%2BA4fzLhPC5hiPlvkdIQiBHpCUcbHWXR9ZNDUdrCRSLFFaOttraRWR1EqJmLw7FOiVj23xuDIxhyMlmJnrwL2bTJCHBYAoKphx6l6kuE0%2BZvwIBRysLTTxCcfGpL8enQJOxtg50Xm1AupAi4lyGOXSuBpu691pfVDg88WqhIriAdUNcdTbEnu3SQ%2BUCYZP%2BX3eICR1Vz2lfnExlrbU5PybKkE190j%2BCAA%3D%3D%3D).

Note how the constant casts are normalized to `0` by `Clang`. However, the non-constant cases remain "native codegen".

This means that if you have `(uint64_t)18446744073709551616.0` you get zero; but it you have `(uint64_t)x` for some `x` that the compiler can't statically determine the value of, then you will get `18446744073709551615` or `0x80000000_00000000`, depending on if you are targeting ARM64 or x64, respectively.

> Looking at the arm64 disassembly, the performance seems to be a problem on x64 only. arm64 can do it in a single instruction. Do you agree?

This depends on what level of consistency we're talking about. If we say everything has to normalize by default, then no. ARM64 does not return `0` by default and so it would likewise no longer be able to do this in a single instruction.

Reply:
I do think it would be "ok" to normalize constant folding. This is no worse off than Clang or the native compilers. It does leave some oddities around whether inlining happens or not and can cause different results across tiering.

I think it would be good to expose additional methods, such as the following. This would allow users who want normalized results to get them.
```csharp
public struct Double
{
    // Repeat for all 12 integer types
    public static bool TryConvertToUInt32(double value, out uint result);
    public static bool TryConvertToUInt64(double value, out ulong result);
}
```

I don't think the inverse is quite as true. That is, I don't think having some `ConvertToUInt64UsingPlatformBehavior` is the same and will cause perf problems, particularly in graphics based applications (like WPF) where floating-point is frequently used for coordinates or colors before being converted to an integer type for use in the UI. It would also measurably impact scenarios like machine learning and would cause issues for users who are looking to trivially vectorize their code (which will continue being important for us and partner teams in the future).

Reply:
Also noting, this only impacts `floating-point to integer` conversions that would overflow in a `checked` context. That is, it only impacts conversions which do not have a well-defined result according to the IEEE 754 specification. The majority of (and normal/expected scenario for) these conversions are well-defined and will always return a correct and consistent result.

Reply:
I have re-read the earlier @davidwrighton's analysis (also forwarded it to you via email). Based on that analysis, I think the plan should be: 
- Fix the runtime on all platforms to use the saturating behavior for simple math (ie match Rust behavior). It will regress some cases and it does not seem to be too bad.
- Work with Roslyn to switch to the same plan so that we have consistent behavior across the platform.

Reply:
I'll give it another readthrough tomorrow.

That would, for ARM64 result in single instruction behavior for `int32`, `int64`, `uint32`, `uint64`. For `int8`, `int16`, `uint8`, and `uint16` it would result in additional cost, the `uint8` case is probably the "most" concerning since `float->byte` is common for Color conversions and the like. It's probably not as bad on `ARM64` since they have a number of saturating APIs.

I'd have to think on if there is something optimal we can do here for x64, but generally speaking this is going to be at least an extra comparison + blend operation at a minimum, per `floating-point to integer` conversion (the inverse case doesn't need special handling, its all well-defined). This will likely have measurable impact, by default, in WPF and ML.NET. I expect it might also impact WinForms, ImageSharp, Paint.NET, and possibly some other apps.

I don't think we can change the default here without provide some way for devs to get the underlying platform behavior (via a single xplat API). We'd likewise need similar APIs for `Vector<T>`, `Vector64<T>`, `Vector128<T>`, and `Vector256<T>`.

Reply:
>  This will likely have measurable impact, by default, in WPF and ML.NET. I expect it might also impact WinForms, ImageSharp, Paint.NET, and possibly some other apps.

Yes, it would be useful to collect some data to see what would be impacted here and how much.

> I don't think we can change the default here without provide some way for devs to get the underlying platform behavior (via a single xplat API). 

Does Rust have an API like that?

Reply:
> Does Rust have an API like that?

Yes, David called it out in the doc that they have some special APIs for this.

Reply:
@jkotas, @davidwrighton. Rust behavior is here: [full link to rust.godbolt](https://rust.godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgATAFZSjgBk8BkwAOXcAI0xiEAA2LlIAB1QFQjsGFzcPbz9k1NsBIJDwliiY%2BMtMawKGIQImYgJM909fSur0uoaCIrDI6LiE827m7Lbhxt6SssGASktUE2Jkdg5EkwiAaioGTbwAZi8AfSIjqkOIBncQPcPZzYBaAFJ97G3DzaeAdgAhJ40AIKbYGbK4sTZMBTvLz/AHfAAisNh6y2O1ux1OVFikku13R92er222M%2Bv1hINB7ghUKxkiRX0RgORG22uzw2JOqDOFzBN3ZkgJLze5y8pL%2BgIpYOp0PpjLhgJRrL2HMx2NxLD52MFRNpYvJIKlkOJdKZDKRCpZaJFnKOBy86puIu1bzteolBqpRrtsvNAMVVsONv5DuhzuVkjdQI94K92J9TItqN21tOJh5eKdjyFmzTou%2B4qjwMNUNz8flfstycDqbVvNDWaJJhJ%2Bf1Rc9JbjprlzKTxqD6Y1xrDrpb7rbMah3q7vv9yZVXODddpw%2BbZLHlIn4bLPaVtJtuZDy4bb1zkcl7Zzh23id386OTZxS61x5zq4L583D%2BvFd7uZtIpDXMwxFM9o2lEVv1nS8MS5WlALuF9dVHQsN3Azs4TNBMfyVB9/wHG4H2Aj5kI/cCr2nLCoNw1VHzxQjELfVtUKNWlZQ4eZaE4HxeE8DgtFIVBOAAJTMAhNgURZlkwT4vH2HhSAITR2PmABrEB9n2AA6DSdN0vTYn0ThJB4pSBM4XgFBADQFKU%2BY4FgJA0BYRI6GichKCclz6BiYguB8WJrJoWgCGiSyIAiUyImCBoAE9OHkpy2EEAB5BhaDivjeCwFhDGAcRMtIfBiEwGw8AAN0wSyCswVQSpMEL4t4YIQs4graDwCJiFilwsFMghiDwDVuHYvgDGABQADU8EwAB3ZLEkYRqZEEEQxHYKRlvkJQ1FM3QEgMIwQFMcx9A6yzIHmVBEhqKqHmSx4nE2epiGATACAeKhMCYAglkwF5ES8H4mDK1QvAsqoSpqBwGGcVwWhyfwYamfoYgATmkPI0gEUZPC4L4khSLGGGR0oBnR9pIc6Z6mjhsY/CsSmBC6SZgj6Um0ekCYaayXH8a5kmZnR%2BYJKWFY9H6zBVh4DiuJMgrBI4Y6CGQTYuE0/zNI0TYIBEk7NlwQgSBkuTSE2FxnNc4hja4WZeEUzLZlU9StL012dIM1rjNIXj%2BIViyrJsh3SHsxAQEWAh1gsCgIE8y29EwfAiBiBJ%2BBW0RxA21OtpUdQColqXeBmrrEkamWOG473TIV5L6sjzZUCoTYlZVtWNa1iBza86JrdtwOtEd0g1I07S3ddwyOC9n3eD9ywA/t/uy7Byv5fMvvlNICriFSexJCAA%3D%3D%3D)

Discussion of the overall issue + the benchmarks they did are here: <https://github.com/rust-lang/rust/issues/10184>

For simple cases like `f32_to_i32` we'd go, on Skylake, from ~7 cycles to between ~19 and ~24 cycles (a theoretical "worst case" taking 3.42x more time).

The measured impact was lower but still significant in some of the areas I called out. For example, there was a measured "x0.82" regression.

Notably, I'm not convinced what `rust` is doing here is the "most efficient" and looking through some of the issues/comments they believe so as well but are somewhat limited by what LLVM can trivially represent today.

We on the other hand, are not, and I believe the right thing here is to do replace the cast with the following in morph:
```csharp
var result = platform_f32_to_i32(value);

if (result == sentinel) // 0x80000000
{
    result = saturating_f32_to_i32(value);
}
```

This should, given branch prediction, help avoid some of the slowdown that occurs in the normal case.

The rust code itself is always dealing with the predictor anyways due to `cmovnp` inserted to handle `nan` (possibly more for a couple of the generated patterns) and so it shouldn't be any slower than the rust code and will be on average smaller.

Reply:
As for the platform specific conversion APIs, they were tracking those via https://github.com/rust-lang/rust/issues/67058.

It looks like they went back and forth between `approx_unchecked` and `to_int_unchecked` before stabilizing on the latter. I think we'll need different names ourselves due to the meaning of `unchecked` in C#.

Reply:
Going to mark this thread as "resolved". I've logged https://github.com/dotnet/runtime/issues/61885 to track the issue more generally and have updated the PR to use saturation.